src/trio/_tests/type_tests/raisesgroup.py

```diff
@@ -0,0 +1,103 @@
+"""The typing of RaisesGroup involves a lot of deception and lies, since AFAIK what we
+actually want to achieve is ~impossible. This is because we specify what we expect with
+instances of RaisesGroup and exception classes, but excinfo.value will be instances of
+[Base]ExceptionGroup and instances of exceptions. So we need to "translate" from
+RaisesGroup to ExceptionGroup.
+
+The way it currently works is that RaisesGroup[E] corresponds to
+ExceptionInfo[BaseExceptionGroup[E]], so the top-level group will be correct. But
+RaisesGroup[RaisesGroup[ValueError]] will become
+ExceptionInfo[BaseExceptionGroup[RaisesGroup[ValueError]]]. To get around that we specify
+RaisesGroup as a subclass of BaseExceptionGroup during type checking - which should mean
+that most static type checking for end users should be mostly correct.
+"""
+from __future__ import annotations
+
+import sys
+from typing import Union
+
+from trio.testing import Matcher, RaisesGroup
+from typing_extensions import assert_type
+
+if sys.version_info < (3, 11):
+    from exceptiongroup import BaseExceptionGroup, ExceptionGroup
+
+# split into functions to isolate the different scopes
+
+
+def check_inheritance_and_assignments() -> None:
+    # Check inheritance
+    _: BaseExceptionGroup[ValueError] = RaisesGroup(ValueError)
+    _ = RaisesGroup(RaisesGroup(ValueError))  # type: ignore
+
+    a: BaseExceptionGroup[BaseExceptionGroup[ValueError]]
+    a = RaisesGroup(RaisesGroup(ValueError))
+    a = BaseExceptionGroup("", (BaseExceptionGroup("", (ValueError(),)),))
```

Comment:
```
  ++ pyright src/trio/_tests/type_tests
    /home/runner/work/trio/trio/src/trio/_tests/type_tests/raisesgroup.py:35:32 - error: Argument type is partially unknown
      Argument corresponds to parameter "args" in function "__init__"
      Argument type is "tuple[BaseExceptionGroup[Unknown]]" (reportUnknownArgumentType)
    /home/runner/work/trio/trio/src/trio/_tests/type_tests/raisesgroup.py:99:13 - error: Argument type is partially unknown
      Argument corresponds to parameter "args" in function "__init__"
      Argument type is "tuple[ExceptionGroup[Unknown]]" (reportUnknownArgumentType)
  2 errors, 0 warnings, 0 informations 
```

Reply:
bug in exceptiongroup: agronholm/exceptiongroup#101